src/jupyterlabcontroller/storage/k8s.py

```diff
@@ -674,3 +682,430 @@ def _standard_metadata(self, name: str) -> V1ObjectMeta:
+        """Remove the set of fileserver objects for a user.  It doesn't
+        return until the objects are no longer present in Kubernetes.
+        """
+        with FILESERVER_LOCK:
+            await self._delete_fileserver_job(username, namespace)
+            await self._delete_fileserver_service(username, namespace)
+            await self._delete_fileserver_gafaelfawringress(
+                username, namespace
+            )
+
+            obj_name = f"{username}-fs"
+
+            for kind in ("gafaelfawringress", "job", "service", "ingress"):
+                await self._wait_for_fileserver_object_deletion(
+                    obj_name=obj_name, namespace=namespace, kind=kind
+                )
+
+    def _get_deletion_read_coro(
+        self, obj_name: str, namespace: str, kind: str
+    ) -> Coroutine[Any, Any, Any]:
+        if kind == "job":
+            return self.batch_api.read_namespaced_job(obj_name, namespace)
+        elif kind == "gafaelfawringress":
+            crd_group = "gafaelfawr.lsst.io"
+            crd_version = "v1alpha1"
+            plural = "gafaelfawringresses"
+            # Note method name inconsistency
+            return self.custom_api.get_namespaced_custom_object(
+                crd_group, crd_version, namespace, plural, obj_name
+            )
+        elif kind == "service":
+            return self.api.read_namespaced_service(obj_name, namespace)
+        elif kind == "ingress":
+            return self.networking_api.read_namespaced_ingress(
+                obj_name, namespace
+            )
+        raise WaitingForObjectError(
+            f"Don't know how to check for {kind} presence."
+        )
+
+    async def _wait_for_fileserver_object_deletion(
+        self, obj_name: str, namespace: str, kind: str
+    ) -> None:
+        """This is as generic as I'm willing to go, and it's probably
+        too generic already."""
+        timeout = 30.0
+        interval = 2.7
+        self._logger.debug(
+            f"Waiting for {kind} deletion",
+            name=obj_name,
+            namespace=namespace,
+        )
+        async with asyncio.timeout(timeout):
+            while True:
+                try:
+                    await self._get_deletion_read_coro(
+                        obj_name, namespace, kind
+                    )
+                except ApiException as e:
+                    if e.status == 404:
+                        return
+                    raise KubernetesError.from_exception(
+                        f"Error waiting for {kind} deletion",
+                        e,
+                        namespace=namespace,
+                        name=obj_name,
+                    ) from e
+                self._logger.debug(
+                    f"{kind} still present; waiting {interval}s "
+                    + "then rechecking."
+                )
+                await asyncio.sleep(interval)
```

Comment:
As mentioned elsewhere, we should prefer event watches to polling.

Reply:
Non-Pod event watches will be address on DM-38972